### PR TITLE
Forsaken story cycle

### PR DESCRIPTION
**Pull request by edmund-h**
## Changelog:
- Forsaken Undercity story cycle now carries over when character dies.
- Characters who grant Lordaeron to a vassal, or characters who are a vassal, will see different events and may experience different outcomes during the Undercity events.
- If you take over the Undercity after someone else founds it, and are of the correct culture, you and they will receive different events on completion.
- Four new random events to give flavor to Forsaken playthroughs.
- Improved performance for undercibty-related checks. 

## Tests:
- [ ] There are no errors in `wc` files in `Documents\Paradox Interactive\Crusader Kings III\logs\error.log` except `portrait_decals.cpp:101`
- [ ] The mod takes less than 5.5 GB in the Task Manager (Windows)

# How to test:

### CODE REVIEW

File: events/wc_events/wc_lordaeron_events.txt [hunk 1]
```diff
@@ -440,4 +633,884 @@ lordaeron.1103 = {
...
+		name = forsaken.9002.opt.a
+		if = { 
+			limit = { scope:new_character.faith = faith:echo_of_life }
+			random_list = { 
+				50 = { 
+					send_interface_toast = {
+						title = forsaken.9002.education_successful
+						left_icon = root
+						custom_tooltip = forsaken.9002.education_successful_tt
+						add_character_modifier = {
```
Comment (helloruhe) [inline events/wc_events/wc_lordaeron_events.txt:1140]:
```suggestion
				trigger = { 
					OR = {
						scope:new_character.faith = faith:alchemical_materialism 
						scope:new_character.faith = faith:forsaken_cult
					}
				}
```

File: events/wc_events/wc_lordaeron_events.txt [hunk 2]
```diff
@@ -440,4 +633,884 @@ lordaeron.1103 = {
...
+					send_interface_toast = {
+						title = forsaken.9002.alterations_failed
+						left_icon = root
+						custom_tooltip = forsaken.9002.alterations_failed_tt
+						random_knight = { 
+							limit = { NOT = { exists = primary_title } }
+							death = {
+								death_reason = death_poison
+							}
+						}
```
Comment (helloruhe) [inline events/wc_events/wc_lordaeron_events.txt:1191]:
```suggestion
				trigger = { 
					OR = {
						scope:new_character.faith = faith:alchemical_materialism 
						scope:new_character.faith = faith:forsaken_cult
					}
				}
```

File: events/wc_events/wc_lordaeron_events.txt [hunk 3]
```diff
@@ -440,4 +633,884 @@ lordaeron.1103 = {
...
+			custom_tooltip = { 
+				text = propaganda_campaign_religious_tooltip
+				start_religious_propaganda_campaign = yes
+			}
+		}
+	}
+}
+
+# neighbor raids 
+forsaken.9004 = {
```
Comment (helloruhe) [inline events/wc_events/wc_lordaeron_events.txt:1347]: going to suggest some changes, dont have do to them but maybe make this option have some chance of the raiding party actually defend against you raiding them back. it makes sense if they have good martial to win but its kinda boring if you win everytime

Comment (edmund-h) [reply events/wc_events/wc_lordaeron_events.txt:1347]: honestly i just got bored making this event lol i might just ice this event tbh

**Review by helloruhe — COMMENTED** [review]
comment ;)

File: events/wc_events/wc_lordaeron_events.txt [hunk 4]
```diff
@@ -440,4 +633,884 @@ lordaeron.1103 = {
...
+				REASON = friend_showed_mercy
+				CHARACTER = root
+				OPINION = 50
+			}
+		}
+		add_hook = {
+			target = scope:new_character
+			type = loyalty_hook
+		}
+	}
```
[the remark this answers is not in the document]
Comment (edmund-h) [reply events/wc_events/wc_lordaeron_events.txt:995]: you just reminded me, i didn't do ANY ai or personality consequences

**Comment by edmund-h** [comment]
going back to a draft to take in Dione's recommendations and add AI values and personality consequences

File: events/wc_events/wc_lordaeron_events.txt [hunk 5]
```diff
@@ -440,4 +633,884 @@ lordaeron.1103 = {
...
+	title = forsaken.9004.title
+	desc = forsaken.9004.desc
+	theme = battle
+
+	right_portrait = { 
+		character = root
+		animation = rage
+	}
+
+	trigger = {
```
[the remark this answers is not in the document]
Comment (edmund-h) [reply events/wc_events/wc_lordaeron_events.txt:1358]: ok but i will put him/her in a less prominent place

File: events/wc_events/wc_lordaeron_events.txt [hunk 6]
```diff
@@ -440,4 +633,884 @@ lordaeron.1103 = {
...
+	}
+
+	# raid them back!
+	option = { 
+		name = forsaken.9004.opt.c
+
+		trigger = { 
+			OR = { 
+				has_trait = vengeful has_trait = wrathful has_trait = brave has_trait = impatient has_trait = irritable 
+				martial > 13
```
[the remark this answers is not in the document]
Comment (edmund-h) [reply events/wc_events/wc_lordaeron_events.txt]: i planned something like this but got too lazy. forgot about the duel mechanic, so thank you!

File: events/wc_events/wc_lordaeron_events.txt [hunk 7]
```diff
@@ -440,4 +633,884 @@ lordaeron.1103 = {
...
+					send_interface_toast = {
+						title = forsaken.9002.education_successful
+						left_icon = root
+						custom_tooltip = forsaken.9002.education_successful_tt
+						add_character_modifier = { 
+							modifier = wc_arcanist_educated_casters
+							years = 2
+						}
+					}
+				}
```
[the remark this answers is not in the document]
Comment (edmund-h) [reply events/wc_events/wc_lordaeron_events.txt:1149]: eh, couldn't hurt. but, only will add it in the success cases

File: common/on_action/wc_lordaeron_on_actions.txt [hunk 8]
```diff
@@ -1,10 +1,39 @@
...
+
+	random_events = { 
+		chance_of_no_event = { 
+			value = 50
+			if = { 
+				limit = { is_of_forsaken_like_culture = no }
+				add = 50
+			}
+		}
+
```
[the remark this answers is not in the document]
Comment (edmund-h) [reply common/on_action/wc_lordaeron_on_actions.txt:32]: oh interesting like if someone else manages to conquer it they can blow it up or something?

[the remark this answers is not in the document]
Comment (helloruhe) [reply common/on_action/wc_lordaeron_on_actions.txt:32]: yeah or if they meet some criteria they can become forsaken themselves... kinda like a spell idk

**Review by helloruhe — CHANGES_REQUESTED** [review]
code lgtm but need to add a cooldown on the events so we dont get the same one everytime.
also is the undercity balanced in the balancing sheet? if it isnt then balance, otherwise thats it

**Review by helloruhe — APPROVED** [review]

**Pull request merged**